[v2] ci: run bm25 and ColBERT test in ci

sam-hey opened the pull request:
cI: Enable skipped tests for ColBERT and BM25
fix: get model meta from CrossEncoder 
ref: use tmp_path for tests
test: add tests for model_meta


## Checklist
<!-- Please do not delete this -->

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 

start in: https://github.com/embeddings-benchmark/mteb/pull/1759#issuecomment-2596042055

KennethEnevoldsen reviewed (commented):
This only really checks if they can be installed. We should add an actual test as well?

sam-hey commented:
Hi @KennethEnevoldsen,

I added two tests in [this PR](https://github.com/embeddings-benchmark/mteb/pull/1759). However, they are not being run in the CI because the dependencies don't exist.

sam-hey commented:
Tests conducted on v2.0.0 with Python 3.11:

checks occasionally take up to 12 minutes.
This branch test runtime: 8-9 minutes.

Performance benchmarks:
ColBERT + BM25: Test execution on my machine completes in 15.17 seconds.

Test coverage summary:
Overall test coverage: 66%.
mteb/models/colbert_models.py: 63%
mteb/models/bm25.py: 94%

sam-hey commented:
I believe I’ve added everything for now and would be happy to get your review of the changes!

- **Skipping pylate**: Skipped for Python 3.9.
- **Updates**:
  - Added extra tests.
  - Introduced `ModelMeta` for CrossEncoders passed as models to `run()`.
- **Pending Decision**: Awaiting your input on skipping BM25s and ColBERT if the version is not available.  

@KennethEnevoldsen

sam-hey commented:
@Samoed can you give me a review if you have some time?

Samoed commented on mteb/models/overview.py:
```diff
@@ -226,6 +231,36 @@ def model_meta_from_hf_hub(model_name: str) -> ModelMeta:
+def model_meta_from_cross_encoder(model: CrossEncoder) -> ModelMeta:
+    try:
+        name = model.model.name_or_path
+        # languages = (
+        #    [model.model_card_data.language]
+        #    if isinstance(model.model_card_data.language, str)
+        #    else model.model_card_data.language
+        # )
```
```suggestion
```

Samoed commented on tests/test_benchmark/test_benchmark_integration_with_datasets.py:
```diff
@@ -18,7 +19,7 @@
     """Test that a task can be fetched and run"""
     eval = MTEB(tasks=[task])
-    eval.run(model, output_folder="tests/results", overwrite_results=True)
+    eval.run(model, output_folder=tmp_path.as_posix(), overwrite_results=True)
```
`str(tmp_path)` can be used there, just to standartize?

sam-hey replied:
I believe .as_posix() is nicer than str(), so I would convert everything to use .as_posix() instead.

Samoed replied:
Yes, I think that too

Samoed commented on tests/test_benchmark/test_benchmark_integration_with_sentencetransformers.py:
```diff
@@ -22,9 +23,11 @@
         model = SentenceTransformer(model_name)
     eval = MTEB(tasks=[task])
-    eval.run(model, output_folder="tests/results", overwrite_results=True)
+    eval.run(model, output_folder=tmp_path.as_posix(), overwrite_results=True)
```
Same

Samoed commented on tests/test_model_meta/test_model_meta.py:
```diff
@@ -0,0 +1,50 @@
+    meta = MTEB.create_model_meta(model)
+
+    assert meta.name == model_name
+    assert meta.revision == model.model_card_data.base_model_revision
```
I think there can be checked more params

sam-hey replied:
I've added some more; please take a look

Samoed replied:
Maybe it would be better to change it like `mteb.get_model("sentence-transformers/average_word_embeddings_levy_dependency")` and `assert meta.name == "sentence-transformers/average_word_embeddings_levy_dependency"`

sam-hey replied:
I actually prefer this approach because:  
- No magic literals.  
- Less error-prone due to typos in the name.

Samoed replied:
But this doesn't check the revision, as the revision can be `None` in both cases, and we're not verifying it.

sam-hey replied:
I added a check to ensure that the revision is the same. The case mentioned above would have worked, but it would have been ignored if there was an error in ST

Samoed commented on tests/test_model_meta/test_model_meta.py:
```diff
@@ -0,0 +1,50 @@
+    output_path = mteb.create_output_folder(
+        model_meta=meta, output_folder=tmp_path.as_posix()
+    )
+    assert Path(output_path).exists()
```
To simplify
```python
output_path = Path(output_path)
assert ...
...
```

Samoed reviewed (commented):
Great!

sam-hey commented:
Why was `MaxSim` renamed to `max_sim`? This change breaks `ColBERT`. @Samoed, I’ll fix it and add a test to prevent similar issues in the future.

Samoed commented:
`max_sim` was used in main branch and I thought it was tested and worked with this name

sam-hey commented:
> `max_sim` was used in main branch and I thought it was tested and worked with this name

The `max_sim` function was updated to `MaxSim` to align with the upcoming pylate integration in a future release.

Samoed commented on mteb/evaluation/MTEB.py:
```diff
@@ -495,7 +498,10 @@ def create_model_meta(model: Encoder) -> ModelMeta:
         else:
             try:
-                meta = model_meta_from_sentence_transformers(model)  # type: ignore
+                if isinstance(model, CrossEncoder):
+                    meta = model_meta_from_cross_encoder(model)
+                else:
+                    meta = model_meta_from_sentence_transformers(model)  # type: ignore
```
I think it would be better to add a `get_model_meta` method and include this `if` condition there.

sam-hey commented on mteb/evaluation/MTEB.py:
```diff
@@ -498,10 +498,7 @@ def create_model_meta(model: Encoder) -> ModelMeta:
-                    meta = model_meta_from_cross_encoder(model)
-                else:
-                    meta = model_meta_from_sentence_transformers(model)  # type: ignore
+                meta = MTEB._get_model_meta(model)
```
I made the method `@staticmethod` because it does not require `self`, making it impossible to call it with `self`.

sam-hey replied:
BTW, the call is also within a method that is static, so `self` is not available.

Samoed commented on tests/test_benchmark/test_models.py:
```diff
@@ -9,9 +12,10 @@
 from .mock_tasks import MockRetrievalTask
 
 
+@pytest.mark.skipif(sys.version_info < (3, 10), reason="Requires Python 3.10 or higher")
```
Let's wait for https://github.com/lightonai/pylate/pull/80 and merge

sam-hey replied:
Great job! However, I'm not sure how long it will take. A new merge doesn't always correspond to a new version being released on PyPI. Since this PR includes a lot of changes, I would recommend proceeding with it and removing the unnecessary parts in a separate PR later on.

Samoed replied:
@sam-hey FYI https://github.com/lightonai/pylate/pull/80#issuecomment-2604790533

sam-hey commented:
Awesome, thanks a lot!
